File: .API/AssetUtil.cs
// Decompiled with JetBrains decompiler
// Type: CloudX.Shared.AssetUtil
// Assembly: CloudX.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 6223B97A-06A5-46CB-9E10-78604961D6EE
// Assembly location: J:\D\SteamLibrary\steamapps\common\NeosVR\HeadlessClient\CloudX.Shared.dll

using System;
using System.IO;
using System.Security.Cryptography;

namespace CloudX.Shared
{
  public static class AssetUtil
  {
    public static int COMPUTE_VERSION
    {
      get
      {
        return 4;
      }
    }

    public static string GenerateHashSignature(string file)
    {
      using (FileStream fileStream = File.OpenRead(file))
        return AssetUtil.GenerateHashSignature((Stream) fileStream);
    }

    public static string GenerateHashSignature(Stream fileStream)
    {
      using (SHA256 shA256 = SHA256.Create())
        return BitConverter.ToString(shA256.ComputeHash(fileStream)).Replace("-", "").ToLower();
    }

    public static Uri GenerateURL(string signature, string extension)
    {
      if (!string.IsNullOrEmpty(extension) && extension[0] != '.')
        extension = "." + extension;
      return new Uri("neosdb:///" + signature + extension);
    }

    public static string ExtractSignature(Uri uri)
    {
      string extension;
      return AssetUtil.ExtractSignature(uri, out extension);
    }

    public static string ExtractSignature(Uri uri, out string extension)
    {
      if (uri.Scheme != "neosdb")
        throw new ArgumentException("Not a NeosDB URI");
      string segment = uri.Segments[1];
      extension = Path.GetExtension(segment);
      return Path.GetFileNameWithoutExtension(segment);
    }

    public static string ComposeIdentifier(string signature, string variant)
    {
      if (string.IsNullOrWhiteSpace(variant))
        return signature;
      return signature + "&" + variant;
    }

    public static void SplitIdentifier(string identifier, out string signature, out string variant)
    {
      int length = identifier.IndexOf('&');
      if (length >= 0)
      {
        variant = identifier.Substring(length + 1);
        signature = identifier.Substring(0, length);
      }
      else
      {
        variant = (string) null;
        signature = identifier;
      }
      signature = signature.ToLower();
    }
  }
}

File: API/AssetUtil.js
const { Out } = require("./Out");
const { Path } = require("./Path");
const { Uri } = require("./Uri");
const SHA256 = require("crypto-js/sha256");
const fs = require("fs");
/**.
 * Utility Class for working with Assets
 *
 * @class AssetUtil
 */
class AssetUtil {
	/**.
	 * Compute Version
	 *
	 * @readonly
	 * @static
	 * @memberof AssetUtil
	 * @returns {4}
	 */
	static get COMPUTE_VERSION() {
		return 4;
	}
	/**.
	 * Generate a SHA256 hash signature for a file
	 *
	 * @static
	 * @param {Buffer | string} file File Stream or Path to file
	 * @memberof AssetUtil
	 * @returns {string} Hash Signature
	 */
	static GenerateHashSignature(file) {
		if (file instanceof String) {
			let fileStream = fs.readFileSync(file);
			return AssetUtil.GenerateHashSignature(fileStream);
		} else {
			return SHA256(file.toString()).toString().replace("-", "").toLowerCase();
		}
	}
	/**.
	 * Generate a NeosDB Asset Url
	 *
	 * @static
	 * @param {string} signature File Signature
	 * @param {string} extension Asset Type
	 * @returns {Uri}
	 * @memberof AssetUtil
	 */
	static GenerateURL(signature, extension) {
		if (!extension.startsWith(".")) extension = "." + extension;
		return new Uri("neosdb:///" + signature + extension);
	}
	/**.
	 * Extract the signature [And Extension] from a URI
	 *
	 * @static
	 * @param {Uri} uri
	 * @param {Out<String>} [extension]
	 * @returns {string} Signature
	 * @memberof AssetUtil
	 */
	static ExtractSignature(uri, extension = new Out()) {
		if (uri.Scheme !== "neosdb") throw new Error("Not a NeosDB URI");
		let segment = uri.Segments[1];
		extension.Out = Path.GetExtension(segment);
		return Path.GetFileNameWithoutExtension(segment);
	}
	/**.
	 * Append a Variant Identifier to a signature
	 *
	 * @static
	 * @param {string} signature
	 * @param {string} variant
	 * @returns {string}
	 * @memberof AssetUtil
	 */
	static ComposeIdentifier(signature, variant) {
		if (String.IsNullOrWhiteSpace(variant)) return signature;
		return signature + "&" + variant;
	}
	/**.
	 * Split the Varient and Signature from a Variant Identifier
	 *
	 * @static
	 * @param {string} identifier
	 * @param {Out<String>} [signature]
	 * @param {Out<String>} [variant]
	 * @memberof AssetUtil
	 */
	static SplitIdentifier(identifier, signature, variant) {
		let length = identifier.indexOf("&");
		if (length >= 0) {
			variant.Out = identifier.substr(length + 1);
			signature.Out = identifier.substr(0, length);
		} else {
			variant.Out = null;
			signature.Out = identifier.toLowerCase();
		}
	}
}
module.exports = {
	AssetUtil,
};
